### PR TITLE
DAOS-8722 dfuse: Add support for dfuse caching readdir replies.

### PR DESCRIPTION
**Pull request by ashleypittman**
Required-githooks: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/object/srv_obj.c [hunk 1]
```diff
@@ -2570,12 +2570,14 @@ obj_tgt_update(struct dtx_leader_handle *dlh, void *arg, int idx,
 		rc = obj_local_rw(exec_arg->rpc, exec_arg->ioc, &dlh->dlh_handle);
 		if (rc != 0)
 			D_CDEBUG(rc == -DER_INPROGRESS || rc == -DER_TX_RESTART ||
-				 (rc == -DER_EXIST && (orw->orw_api_flags &
-				  (DAOS_COND_DKEY_INSERT | DAOS_COND_AKEY_INSERT))) ||
-				 (rc == -DER_NONEXIST && (orw->orw_api_flags &
-				  (DAOS_COND_DKEY_UPDATE | DAOS_COND_AKEY_UPDATE))),
-				 DB_IO, DLOG_ERR,
-				 DF_UOID": error="DF_RC".\n", DP_UOID(orw->orw_oid), DP_RC(rc));
+				     (rc == -DER_EXIST &&
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/object/srv_obj.c:2573]: Interesting, this was formatted via the commit hook so should be correct, I'll fix on next push.

File: utils/node_local_test.py [hunk 2]
```diff
@@ -1059,24 +1059,24 @@ def get_cmd_prefix(self):
                                          suffix='.memcheck', delete=False) as log_file:
             self._xml_file = log_file.name
 
-        cmd = ['valgrind', '--fair-sched=yes']
+        cmd = ['valgrind',
+               f'--xml-file={self._xml_file}',
+               '--xml=yes',
+               '--fair-sched=yes',
+               '--gen-suppressions=all',
+               '--error-exitcode=42']
```
Comment (ashleypittman) [inline utils/node_local_test.py:1067]: This change is not required for this PR and is not a functional change however moves around the command line options to valgrind so that 'ps' can show the logfile name and therefore the running test.

**Comment by ashleypittman** [comment]
Can I request reviews on this please.  It's been ready for review and passing all tests for over a week, I've just re-merged with master as the queues are short so it's up to date.

File: utils/node_local_test.py [hunk 3]
```diff
@@ -1957,7 +1976,40 @@ def test_readdir_hard(self):
...
+        files2 = []
+        with os.scandir(test_dir) as entries:
+            with os.scandir(test_dir) as second:
+                files2.append(next(second).name)
+                for entry in entries:
+                    files.append(entry.name)
+                    assert len(files) < count + 2
+                for entry in second:
+                    files2.append(entry.name)
+                    assert len(files2) < count + 2
```
Comment (daltonbohning) [inline utils/node_local_test.py:1994]: FYI - If you wanted to interleave the two:
```
for pair in itertools.zip_longest(entries, second):
    files.append(pair[0])
    files2.append(pair[1])
```

File: utils/node_local_test.py [hunk 4]
```diff
@@ -1957,7 +1976,40 @@ def test_readdir_hard(self):
...
+
+        print('Reads are from list 2, 1, 1, 2.')
+        print(files)
+        print(files2)
+        assert files == files2
+        assert len(files) == count
```
Comment (daltonbohning) [inline utils/node_local_test.py:2000]: For robustness, should we save the file names from create and verify readdir returns the correct file names?

File: src/client/dfuse/ops/readdir.c [hunk 5]
```diff
@@ -207,79 +280,272 @@ dfuse_readdir_reset(struct dfuse_readdir_hdl *hdl)
...
+	 * however readdir does get called concurrently with releasedir for the same inode so
+	 * protect this section with a spinlock.
+	 */
+	D_SPIN_LOCK(&fs_handle->dpi_info->di_lock);
 
-	if (offset == READDIR_EOD) {
-		oh->doh_kreaddir_finished = true;
-		DFUSE_TRA_DEBUG(oh, "End of directory %#lx", offset);
+	if (oh->doh_rd == NULL) {
+		if (oh->doh_ie->ie_rd_hdl &&
```
Comment (daltonbohning) [inline src/client/dfuse/ops/readdir.c:303]: To be clear, we need to lock whether oh->doh_rd is NULL or not?

File: utils/node_local_test.py (hunk 4 above)
Comment (ashleypittman) [reply utils/node_local_test.py:2000]: We could but that part is extremely likely to go wrong.

File: src/client/dfuse/ops/readdir.c (hunk 5 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/readdir.c:303]: Good question.  Superficially it would seem not.  Will take a close look tomorrow.

File: utils/node_local_test.py (hunk 4 above)
Comment (daltonbohning) [reply utils/node_local_test.py:2000]: But shouldn't it work? If readdir returns the wrong file names, it's functionally not working

**Review by jolivier23 — APPROVED** [review]

**Review by mchaarawi — CHANGES_REQUESTED** [review]
before we review/land, i'd like to understand the need for this if kernel readdir caching already is supposed to work.
the results from resnet50 have not proved that this PR helps the initial parallel traversal of the training directory either.

**Comment by ashleypittman** [comment]
The resent issues should be fixed now and were an issue with this PR, an uninitialised stack varlaible was being passed to dfs_lookup_x.

Please review.

**Review by daltonbohning — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]
unfortunately the code has become hard to follow. it would be great to add a readme.md for how readdir is implemented in dfuse. i don't see anything wrong in the patch though to block it from landing given that it now gives the expected result.

the PR description is also empty.

**Comment by ashleypittman** [comment]
PR description updated with an appropriate commit message, I will submit a new PR to contain a description of the internals.
